tests/stdlib/tos.nim

```diff
@@ -333,3 +365,102 @@ block ospaths:
   doAssert joinPath("", "lib") == "lib"
   doAssert joinPath("", "/lib") == unixToNativePath"/lib"
   doAssert joinPath("usr/", "/lib") == unixToNativePath"usr/lib"
+
+block parentDir:
+  let examples = [
+    ("/usr/local/bin", "/usr/local"),
+    ("foo/bar.nim", "foo"),
+    ("foo//bar.nim", "foo"),
+    ("foo//bar//", "foo"),
+    ("foo///bar", "foo"),
+    ("foo///bar/.", "foo"),
+    ("./foo///bar", "./foo"),
+    (".//foo///bar", ".//foo"),
+    ("/.//foo///bar", "/.//foo"),
+    ("foo/bar//a/./.", "foo/bar"),
+    ("/a/bar", "/a"),
+    ("/bar", "/"),
+
+    # same as in shell, `cd ..` returns `.` when pwd = "foo"; "." is not same as empty
+    # see https://github.com/nim-lang/Nim/pull/10018#issuecomment-447996816
+    ("a/./.", "."),
+    ("./bar", "."),
+    (".//bar", "."),
+    ("bar", "."),
+
+    (".git", "."),
+    (".git.bak1", "."),
+
+    # absolute remains absolute
+    ("/", "/"),
+    ("/.", "/"),
+    ("/..", "/"),
+    ("/./", "/"),
+
+    # return empty when no parent possible
+    ("", ""),
+    (".", ""),
+    ("./", ""),
+    ("..", ""),
+    ("../", ""),
+    ("../..", ""),
+
+    # regression tests
+
+    # fix #8734 (bug 2)
+    ("a/b//", "a"),
+    ("a/b/", "a"),
+
+    # fix #8734 (bug 3)
+    ("/", "/"),
+
+    # fix #8734 (bug 4)
+    ("/a.txt", "/"),
+  ]
+
+  doAssert runTestCases("parentDir", examples, parentDir(it))
+
+import sequtils
+
+block tailDir:
+  let examples = [
+    ("/usr/local/bin", "usr/local/bin"),
+    ("usr/local/bin", "local/bin"),
+
+    # issue #8395; todo: fix
+    # ("//usr//local//bin//", "usr//local//bin//"),
+    # ("usr//local//bin//", "local/bin//"),
+  ]
+
+  doAssert runTestCases("tailDir", examples, tailDir(it))
+
+block parentDirs:
+  template test(iter: untyped, expected: seq[string]): untyped =
+    let lhs = toSeq(iter)
+    let rhs = expected.mapIt(it.unixToNativePath)
+    assertEquals lhs, rhs
+
+  # fromRoot=false, inclusive=true
+  test parentDirs("a/b/c".unixToNativePath), @["a/b/c", "a/b", "a"]
+  test parentDirs("/a/b/c".unixToNativePath), @["/a/b/c", "/a/b", "/a", "/"]
+  test parentDirs("//a/b//c//".unixToNativePath), @["//a/b//c", "//a/b", "//a", "/"]
```

Comment:
Why doesn't parentDirs normalize the paths?

Reply:
same behavior as in most (all?) other languages: it normalizes only the endpoint:

```nim
os.path.dirname("//abc//def//")
Out[3]: '//abc//def'
```

D:
```nim
rdmd --eval 'writeln(dirName("//abc//def//"))'
//abc
```

bash:
```
dirname //abc//def//
//abc
```

etc...

Reply:
Sounds like a bug to me, but ok.